uefi/src/allocator.rs

```diff
@@ -100,7 +102,8 @@ unsafe impl GlobalAlloc for Allocator {
             // `allocate_pool` always provides eight-byte alignment, so we can
             // use `allocate_pool` directly.
             boot::allocate_pool(memory_type, size)
-                .map(|ptr| ptr.as_ptr())
+                .map(|mut ptr: NonNull<[u8]>| unsafe { ptr.as_mut() })
```

Comment:
I think this is unsound: `as_mut` creates a reference, but that's not valid because the allocated memory is uninitialized. I think instead this can be:
```rust
            boot::allocate_pool(memory_type, size)
                .map(|ptr: NonNull<[u8]>| ptr.as_ptr().cast::<u8>())
                .unwrap_or(ptr::null_mut())
```

(Potentially in the future it could be further simplified to `ptr.as_mut_ptr()`, but [it's not yet stable](https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.as_mut_ptr).)

Reply:
> I think this is unsound: as_mut creates a reference, but that's not valid because the allocated memory is uninitialized.

Hm.... ( I think we had a similar discussion before but I cannot recall details) allocating a `NonNull<[u8]>` is something that eventually everyone using the `allocator_api` is doing. As long as we handle it as `[u8]`, I think any bit pattern can be safely considered as initialized.

I think I've read in a Ralf Jung et. al. discussion that it is not yet decided whether every bit pattern for u8 can indeed be safe and considered as initialized.

Or in other words: How do we can get a `&[u8]` from an `NonNull<[u8]>` in a safe way without zeroing all data first or write any other data to each byte?